### ISSUE TITLE
No subject provided

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2206]
Migrated from [rt.perl.org#3505](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3505) (status was 'resolved')

Searchable as RT3505$

**Comment by p5pRT** [comment]
### From @simoncozens



This should be a nice easy job for someone\.

&nbsp;    $a=v120\.300; $b = v200\.400;
&nbsp;    $a^=$a;

This'll put illegal UTF8 characters in $a\, and here's why\.

Take a look at Perl\_do\_vop in doop\.c of cfgperl\.

We're reading from the PV in left\, which we store in lc\, and the PV in
right\, which we store in rc\, and we're or'ing each character\. We put the
result in the PV of the target\, sv\, which we call dc\.

With me so far? So here's what happens&#8203;:

&nbsp;    120 ^ 200 = 176
&nbsp;    200 ^ 300 = 188

So\, the first character of dc becomes 176\, but we're in Unicode now\, so
that encodes as \\302\\260\. You'll note this is two bytes\.

Here's the kicker&#8203;: Since the target is also the left\-hand side of the
op\, \($a = $a^$b\) we've just written two bytes to the start of the
string\. And since the original character\, 120\, was only one byte\, we've
scribbled over the start of the next character\!

To fix this&#8203;: if \(sv==left | sv==right\)
&nbsp;    \* Keep dc as a new\, separate string until we've finished
&nbsp;    \* Set SvPV\(sv\) to dc when we're done
&nbsp;    \* Free whichever of lc or rc it replaces\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.0:

Configured by root at Thu Apr 13 08:15:45 JST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.3.99-pre3, archname=i686-linux
    uname='linux othersideofthe.earth.li 2.3.99-pre3 #3 smp sat mar 25 23:54:38 jst 2000 i686 unknown '
    config_args='-d'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O2', gccversion=2.95.2 20000313 (Debian GNU/Linux)
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt
    libc=/lib/libc-2.1.3.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/local/lib/perl5/5.6.0/i686-linux
    /usr/local/lib/perl5/5.6.0
    /usr/local/lib/perl5/site_perl/5.6.0/i686-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl/5.5.670
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/simon
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/simon/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/rhs/bin:/usr/local/bin:/usr/local/sbin:/usr/sbin:/sbin:/opt/kde/bin:/home/simon/bin
    PERL_BADLANG (unset)
    SHELL=/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

In message \<20000714151309\.7170\.qmail@&#8203;othersideofthe\.earth\.li>
&nbsp;          "Simon Cozens" \<simon@&#8203;othersideofthe\.earth\.li> wrote&#8203;:

> To fix this&#8203;: if \(sv==left | sv==right\)
>     \* Keep dc as a new\, separate string until we've finished
>     \* Set SvPV\(sv\) to dc when we're done
>     \* Free whichever of lc or rc it replaces\.

Something like this you mean&#8203;:

diff \-c 'perl\-5\.6\.0/doop\.c' 'perl\-20000714\-002/doop\.c'
Index&#8203;: \./doop\.c
\*\*\* \./doop\.c    Sun Mar 12 03&#8203;:36&#8203;:32 2000
\-\-\- \./doop\.c    Fri Jul 14 23&#8203;:56&#8203;:54 2000
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\*\*\* 1062\,1067 \*\*\*\*
\-\-\- 1062\,1068 \-\-\-\-
&nbsp;      char \*rsave;
&nbsp;      bool left\_utf = DO\_UTF8\(left\);
&nbsp;      bool right\_utf = DO\_UTF8\(right\);
\+     I32 needlen;

&nbsp;      if \(left\_utf && \!right\_utf\)
&nbsp;      sv\_utf8\_upgrade\(right\);
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\*\*\* 1074\,1090 \*\*\*\*
&nbsp;      rsave = rc = SvPV\(right\, rightlen\);
&nbsp;      len = leftlen \< rightlen ? leftlen : rightlen;
&nbsp;      lensave = len;
\!     if \(SvOK\(sv\) || SvTYPE\(sv\) > SVt\_PVMG\) \{
&nbsp;      STRLEN n\_a;
&nbsp;      dc = SvPV\_force\(sv\, n\_a\);
&nbsp;      if \(SvCUR\(sv\) \< len\) \{
&nbsp;          dc = SvGROW\(sv\, len \+ 1\);
&nbsp;          \(void\)memzero\(dc \+ SvCUR\(sv\)\, len \- SvCUR\(sv\) \+ 1\);
&nbsp;      \}
&nbsp;      \}
&nbsp;      else \{
\!     I32 needlen = \(\(optype == OP\_BIT\_AND\)
\!             ? len : \(leftlen > rightlen ? leftlen : rightlen\)\);
&nbsp;      Newz\(801\, dc\, needlen \+ 1\, char\);
&nbsp;      \(void\)sv\_usepvn\(sv\, dc\, needlen\);
&nbsp;      dc = SvPVX\(sv\);        /\* sv\_usepvn\(\) calls Renew\(\) \*/
\-\-\- 1075\,1097 \-\-\-\-
&nbsp;      rsave = rc = SvPV\(right\, rightlen\);
&nbsp;      len = leftlen \< rightlen ? leftlen : rightlen;
&nbsp;      lensave = len;
\!     if \(\(left\_utf || right\_utf\) && \(sv == left || sv == right\)\) \{
\!     needlen = optype == OP\_BIT\_AND ? len : leftlen \+ rightlen;
\!     Newz\(801\, dc\, needlen \+ 1\, char\);
\!     \}
\!     else if \(SvOK\(sv\) || SvTYPE\(sv\) > SVt\_PVMG\) \{
&nbsp;      STRLEN n\_a;
&nbsp;      dc = SvPV\_force\(sv\, n\_a\);
&nbsp;      if \(SvCUR\(sv\) \< len\) \{
&nbsp;          dc = SvGROW\(sv\, len \+ 1\);
&nbsp;          \(void\)memzero\(dc \+ SvCUR\(sv\)\, len \- SvCUR\(sv\) \+ 1\);
&nbsp;      \}
\+     if \(optype \!= OP\_BIT\_AND && \(left\_utf || right\_utf\)\)
\+         dc = SvGROW\(sv\, leftlen \+ rightlen \+ 1\);
&nbsp;      \}
&nbsp;      else \{
\!     needlen = \(\(optype == OP\_BIT\_AND\)
\!             ? len : \(leftlen > rightlen ? leftlen : rightlen\)\);
&nbsp;      Newz\(801\, dc\, needlen \+ 1\, char\);
&nbsp;      \(void\)sv\_usepvn\(sv\, dc\, needlen\);
&nbsp;      dc = SvPVX\(sv\);        /\* sv\_usepvn\(\) calls Renew\(\) \*/
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\*\*\* 1093\,1106 \*\*\*\*
&nbsp;      \(void\)SvPOK\_only\(sv\);
&nbsp;      if \(left\_utf || right\_utf\) \{
&nbsp;      UV duc\, luc\, ruc;
&nbsp;      STRLEN lulen = leftlen;
&nbsp;      STRLEN rulen = rightlen;
\-     STRLEN dulen = 0;
&nbsp;      I32 ulen;

\-     if \(optype \!= OP\_BIT\_AND\)
\-         dc = SvGROW\(sv\, leftlen\+rightlen\+1\);
\-
&nbsp;      switch \(optype\) \{
&nbsp;      case OP\_BIT\_AND&#8203;:
&nbsp;          while \(lulen && rulen\) \{
\-\-\- 1100\,1110 \-\-\-\-
&nbsp;      \(void\)SvPOK\_only\(sv\);
&nbsp;      if \(left\_utf || right\_utf\) \{
&nbsp;      UV duc\, luc\, ruc;
\+     char \*dcsave = dc;
&nbsp;      STRLEN lulen = leftlen;
&nbsp;      STRLEN rulen = rightlen;
&nbsp;      I32 ulen;

&nbsp;      switch \(optype\) \{
&nbsp;      case OP\_BIT\_AND&#8203;:
&nbsp;          while \(lulen && rulen\) \{
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\*\*\* 1113\,1120 \*\*\*\*
&nbsp;          duc = luc & ruc;
&nbsp;          dc = \(char\*\)uv\_to\_utf8\(\(U8\*\)dc\, duc\);
&nbsp;          \}
\!         dulen = dc \- SvPVX\(sv\);
\!         SvCUR\_set\(sv\, dulen\);
&nbsp;          break;
&nbsp;      case OP\_BIT\_XOR&#8203;:
&nbsp;          while \(lulen && rulen\) \{
\-\-\- 1117\,1125 \-\-\-\-
&nbsp;          duc = luc & ruc;
&nbsp;          dc = \(char\*\)uv\_to\_utf8\(\(U8\*\)dc\, duc\);
&nbsp;          \}
\!         if \(sv == left || sv == right\)
\!         \(void\)sv\_usepvn\(sv\, dcsave\, needlen\);
\!         SvCUR\_set\(sv\, dc \- dcsave\);
&nbsp;          break;
&nbsp;      case OP\_BIT\_XOR&#8203;:
&nbsp;          while \(lulen && rulen\) \{
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\*\*\* 1140\,1147 \*\*\*\*
&nbsp;          dc = \(char\*\)uv\_to\_utf8\(\(U8\*\)dc\, duc\);
&nbsp;          \}
&nbsp;        mop\_up\_utf&#8203;:
\!         dulen = dc \- SvPVX\(sv\);
\!         SvCUR\_set\(sv\, dulen\);
&nbsp;          if \(rulen\)
&nbsp;          sv\_catpvn\(sv\, rc\, rulen\);
&nbsp;          else if \(lulen\)
\-\-\- 1145\,1153 \-\-\-\-
&nbsp;          dc = \(char\*\)uv\_to\_utf8\(\(U8\*\)dc\, duc\);
&nbsp;          \}
&nbsp;        mop\_up\_utf&#8203;:
\!         if \(sv == left || sv == right\)
\!         \(void\)sv\_usepvn\(sv\, dcsave\, needlen\);
\!         SvCUR\_set\(sv\, dc \- dcsave\);
&nbsp;          if \(rulen\)
&nbsp;          sv\_catpvn\(sv\, rc\, rulen\);
&nbsp;          else if \(lulen\)
diff \-c 'perl\-5\.6\.0/t/op/bop\.t' 'perl\-20000714\-002/t/op/bop\.t'
Index&#8203;: \./t/op/bop\.t
\*\*\* \./t/op/bop\.t    Mon Feb 28 09&#8203;:38&#8203;:33 2000
\-\-\- \./t/op/bop\.t    Sat Jul 15 00&#8203;:07&#8203;:03 2000
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\*\*\* 9\,15 \*\*\*\*
&nbsp;      unshift @&#8203;INC\, '\.\./lib';
&nbsp;  \}

\! print "1\.\.30\\n";

&nbsp;  \# numerics
&nbsp;  print \(\(0xdead & 0xbeef\) == 0x9ead ? "ok 1\\n" : "not ok 1\\n"\);
\-\-\- 9\,15 \-\-\-\-
&nbsp;      unshift @&#8203;INC\, '\.\./lib';
&nbsp;  \}

\! print "1\.\.35\\n";

&nbsp;  \# numerics
&nbsp;  print \(\(0xdead & 0xbeef\) == 0x9ead ? "ok 1\\n" : "not ok 1\\n"\);
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\*\*\* 81\,83 \*\*\*\*
\-\-\- 81\,98 \-\-\-\-
&nbsp;  print "ok 28\\n" if sprintf\("%vd"\, v4095\.801\.4095 & v801\.4095\) eq '801\.801';
&nbsp;  print "ok 29\\n" if sprintf\("%vd"\, v4095\.801\.4095 | v801\.4095\) eq '4095\.4095\.4095';
&nbsp;  print "ok 30\\n" if sprintf\("%vd"\, v801\.4095 ^ v4095\.801\.4095\) eq '3294\.3294\.4095';
\+
\+ \#
\+ print "ok 31\\n" if sprintf\("%vd"\, v120\.v300 & v200\.v400\) eq '72\.256';
\+ print "ok 32\\n" if sprintf\("%vd"\, v120\.v300 | v200\.v400\) eq '248\.444';
\+ print "ok 33\\n" if sprintf\("%vd"\, v120\.v300 ^ v200\.v400\) eq '176\.188';
\+
\+ \#
\+ my $a = v120\.300;
\+ my $b = v200\.400;
\+ $a ^= $b;
\+ print "ok 34\\n" if sprintf\("%vd"\, $a\) eq '176\.188';
\+ my $a = v120\.300;
\+ my $b = v200\.400;
\+ $a |= $b;
\+ print "ok 35\\n" if sprintf\("%vd"\, $a\) eq '248\.444';

Tom

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Tom Hughes \(lists\.p5p\)&#8203;:
>In message \<20000714151309\.7170\.qmail@&#8203;othersideofthe\.earth\.li>
>          "Simon Cozens" \<simon@&#8203;othersideofthe\.earth\.li> wrote&#8203;:
>
>> To fix this&#8203;: if \(sv==left | sv==right\)
>>     \* Keep dc as a new\, separate string until we've finished
>>     \* Set SvPV\(sv\) to dc when we're done
>>     \* Free whichever of lc or rc it replaces\.
>
>Something like this you mean&#8203;:

A bit like that\, yes\. Thanks\!

Bug\-people\, close that ticket\!

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

In message \<slrn8n07i0\.viv\.simon@&#8203;justanother\.perlhacker\.org>
&nbsp;          simon@&#8203;brecon\.co\.uk \(Simon Cozens\) wrote&#8203;:

> Tom Hughes \(lists\.p5p\)&#8203;:
>
> >Something like this you mean&#8203;:
>
> A bit like that\, yes\. Thanks\!
>
> Bug\-people\, close that ticket\!

Done\.

Tom

**Comment by p5pRT** [comment]
### From @jhi

On Sat\, Jul 15\, 2000 at 12&#8203;:21&#8203;:56AM \+0100\, Tom Hughes wrote&#8203;:
> In message \<20000714151309\.7170\.qmail@&#8203;othersideofthe\.earth\.li>
>           "Simon Cozens" \<simon@&#8203;othersideofthe\.earth\.li> wrote&#8203;:
> 
> > To fix this&#8203;: if \(sv==left | sv==right\)
> >     \* Keep dc as a new\, separate string until we've finished
> >     \* Set SvPV\(sv\) to dc when we're done
> >     \* Free whichever of lc or rc it replaces\.
> 
> Something like this you mean&#8203;:

In it is\.
